drivers/clock_control/clock_control_nrf2_global_hsfll.c

```diff
@@ -0,0 +1,291 @@
+		return "GDFS_FREQ_HIGH";
+	case GDFS_FREQ_MEDHIGH:
+		return "GDFS_FREQ_MEDHIGH";
+	case GDFS_FREQ_MEDLOW:
+		return "GDFS_FREQ_MEDLOW";
+	case GDFS_FREQ_LOW:
+		return "GDFS_FREQ_LOW";
+	default:
+		break;
+	}
+
+	return "UNKNOWN";
+}
+
+static void global_hsfll_work_handler(struct k_work *work)
+{
+	struct global_hsfll_dev_data *dev_data =
+		CONTAINER_OF(work, struct global_hsfll_dev_data, clk_cfg.work);
+	const struct device *dev = dev_data->dev;
+	uint8_t freq_idx;
+	enum gdfs_frequency_setting target_freq;
+	nrfs_err_t err;
+
+	freq_idx = clock_config_update_begin(work);
+	target_freq = global_hsfll_freq_idx_to_nrfs_freq(dev, freq_idx);
+
+	LOG_DBG("requesting %s", global_hsfll_gdfs_freq_to_str(target_freq));
+	err = nrfs_gdfs_request_freq(target_freq, dev_data);
+	if (err != NRFS_SUCCESS) {
+		clock_config_update_end(&dev_data->clk_cfg, -EIO);
+		return;
+	}
+
+	k_work_schedule(&dev_data->timeout_dwork, GLOBAL_HSFLL_FREQ_REQ_TIMEOUT);
+}
+
+static void global_hsfll_evt_handler(struct k_work *work)
+{
+	struct global_hsfll_dev_data *dev_data =
+		CONTAINER_OF(work, struct global_hsfll_dev_data, evt_work);
+	int rc;
+
+	k_work_cancel_delayable(&dev_data->timeout_dwork);
+	rc = dev_data->evt == NRFS_GDFS_EVT_FREQ_CONFIRMED ? 0 : -EIO;
+	clock_config_update_end(&dev_data->clk_cfg, rc);
+}
+
+static void global_hfsll_nrfs_gdfs_init_evt_handler(nrfs_gdfs_evt_t const *p_evt, void *context)
+{
+	struct global_hsfll_dev_data *dev_data = context;
+
+	dev_data->evt = p_evt->type;
+	k_sem_give(&dev_data->evt_sem);
+}
+
+static void global_hfsll_nrfs_gdfs_evt_handler(nrfs_gdfs_evt_t const *p_evt, void *context)
+{
+	struct global_hsfll_dev_data *dev_data = context;
+
+	if (k_work_is_pending(&dev_data->evt_work)) {
+		return;
+	}
+
+	dev_data->evt = p_evt->type;
+	k_work_submit(&dev_data->evt_work);
+}
+
+static void global_hsfll_timeout_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct global_hsfll_dev_data *dev_data =
+		CONTAINER_OF(dwork, struct global_hsfll_dev_data, timeout_dwork);
+
+	clock_config_update_end(&dev_data->clk_cfg, -ETIMEDOUT);
+}
+
+static int global_hfsll_init(const struct device *dev)
+{
+	struct global_hsfll_dev_data *dev_data = dev->data;
+	nrfs_err_t err;
+	int rc;
+
+	k_sem_init(&dev_data->evt_sem, 0, 1);
+	k_work_init_delayable(&dev_data->timeout_dwork, global_hsfll_timeout_handler);
+	k_work_init(&dev_data->evt_work, global_hsfll_evt_handler);
+
+	err = nrfs_gdfs_init(global_hfsll_nrfs_gdfs_init_evt_handler);
+	if (err != NRFS_SUCCESS) {
+		return -EIO;
+	}
+
+	LOG_DBG("initial request %s", global_hsfll_gdfs_freq_to_str(GDFS_FREQ_LOW));
+	err = nrfs_gdfs_request_freq(GDFS_FREQ_LOW, dev_data);
+	if (err != NRFS_SUCCESS) {
+		return -EIO;
+	}
+
+	rc = k_sem_take(&dev_data->evt_sem, GLOBAL_HSFLL_FREQ_REQ_TIMEOUT);
+	if (rc) {
+		return -EIO;
+	}
+
+	if (dev_data->evt != NRFS_GDFS_EVT_FREQ_CONFIRMED) {
+		return -EIO;
+	}
+
+	nrfs_gdfs_uninit();
```

Comment:
Why do you uninit here ?

Reply:
I guess its to have the same state after the driver has initialized regardless off whether or not you have enabled `GLOBAL_HSFLL_INIT_LOW_REQ` ?

Reply:
ok I see this is when GLOBAL_HSFLL_INIT_LOW_REQ is enabled